pkg/querier/distributor_queryable.go

```diff
@@ -198,3 +200,48 @@ func (q *distributorQuerier) LabelNames() ([]string, storage.Warnings, error) {
 func (q *distributorQuerier) Close() error {
 	return nil
 }
+
+type distributorExemplarQueryable struct {
+	distributor Distributor
+}
+
+func newDistributorExemplarQueryable(d Distributor) storage.ExemplarQueryable {
+	return &distributorExemplarQueryable{
+		distributor: d,
+	}
+}
+
+func (d distributorExemplarQueryable) ExemplarQuerier(ctx context.Context) (storage.ExemplarQuerier, error) {
+	return &distributorExemplarQuerier{
+		distributor: d.distributor,
+		ctx:         ctx,
+	}, nil
+}
+
+type distributorExemplarQuerier struct {
+	distributor Distributor
+	ctx         context.Context
+}
+
+// Select querys for exemplars, prometheus' storage.ExemplarQuerier's Select function takes the time range as two int64 values.
+func (q *distributorExemplarQuerier) Select(start, end int64, matchers ...[]*labels.Matcher) ([]exemplar.QueryResult, error) {
+	allResults, err := q.distributor.QueryExemplars(q.ctx, model.Time(start), model.Time(end), matchers...)
+
+	if err != nil {
+		return nil, err
+	}
+
+	var e exemplar.QueryResult
+	ret := make([]exemplar.QueryResult, len(allResults.Timeseries))
+	for _, ts := range allResults.Timeseries {
+		// TODO: (callum) track down why we see empty exemplars slices here
+		// but not in the distributor/ingester code.
```

Comment:
Can you elaborate on this? This looks something worth to investigate before merging this PR.

Reply:
When I first deployed this branch to one of our Cortex clusters I noticed intermittent query failures in Grafana itself, and looking at the query response I saw a lot of results (they always seemed to be at the beginning of the overall `[]exemplar.QueryResult` slice) that had empty series labels and 0 len exemplar slices. The `if` condition right below was my quick fix. 

Yesterday I deployed these changes with additional debug logging, and this block was the only place I saw empty results. Which doesn't really make sense to me. This TODO is a note to me to look into this more. If there's a way to mock a query result from an Ingester replica set then I could try and reproduce in a test, didn't see anything obvious.

Reply:
Fixed the issue in the latest commit(s).